folding/base/validator.py

```diff
@@ -137,10 +138,28 @@ def run(self):
         # This loop maintains the validator's operations until intentionally stopped.
         try:
             while True:
+
+                # Our BaseValidator logic is intentionally as generic as possible so that the Validator neuron can apply problem-specific logic
                 bt.logging.info(f"step({self.step}) block({self.block})")
 
-                # Run multiple forwards concurrently.
-                self.loop.run_until_complete(self.concurrent_forward())
+                # Check if we need to add more jobs to the queue
+                if self.store.get_queue(ready=False) < self.config.queue_size:
+                    # Here is where we select, download and preprocess a pdb
+                    # We also assign the pdb to a group of workers (miners), based on their workloads
+                    self.add_jobs(k = self.config.queue_size - len(queue))
+                    continue
+
+                # Check if we have any ready jobs in the queue
+                queue = self.store.get_queue()
+
+                # TODO: maybe concurrency for the loop below
+                for job in queue.queue:
+                    # Here we straightforwardly query the workers associated with each job and update the jobs accordingly
+                    event = self.forward(job)
```

Comment:
I also understand this now, since we are calling `add_jobs` that calls `create_new_challenge` in the validator base class.  However, I don't see how this works as an async clock. The `self.forward` method you're calling here is an async forward that calls `await run_step`. I do not see any communication logic that only submits jobs to the miner and walks away, but rather we are still waiting for the miners to return data on their own clock (when the simulation finishes) since we are using `await dendrite`. 

Is there some logic we have yet to implement? This is the main crux (haha nice) of my miner development right now, since I am unsure so to what the validator expects to receive from the miner, and if this is purely a validator-ping relationship

Reply:
> it doesn't look like you have altered the validators current forward method to return anything, or use the `create_new_challenge` or `try_prepare_challenge`. am i missing something?

Yes I don’t use the forward method any more. Didn’t seem necessary since the network request logic is pretty simple now